hashicorp/terraform-provider-tfe
# feat(workspace): add auto destroy attribute to resource and data source

## Description

This change adds `auto_destroy_at` to Workspace resource and data source schemas. The Workspaces API accepts three possible types of values for auto_destroy_at:

- a set value
  - e.g. a payload with the shape `{ "data": { "attributes": { "name": "some-workspace", "auto-destroy-at": "2006-01-02T15:04:05Z" }}}`
- an ignored value
  - e.g. a payload with the shape `{ "data": { "attributes": { "name": "some-workspace" }}}`
- and an unset value
  - e.g. a payload with the shape `{ "data": { "attributes": { "name": "some-workspace", "auto-destroy-at": null }}}`

_Remember to:_


## Testing plan

(these follow the integration tests added as part of these changes)
1.  create a workspace with `auto_destroy_at` unset
    ```
    resource "tfe_workspace" "foobar" {
      name            = "workspace-test-0"
      organization    = "hashicorp"
    }
    ```
1. confirm no diffs for future plans
1. update the workspace with a set `auto_destroy_at`
    ```
    resource "tfe_workspace" "foobar" {
      name            = "workspace-test-0"
      organization    = "hashicorp"
      auto_destroy_at = "2024-05-31T14:04:00Z"
    }
    ```
1.  confirm that `auto_destroy_at` is set in state and that no further diff occurs with future plans
1. repeat above steps but creating workspace with `auto_destroy_at` then removing in the update

## External links

_Include any links here that might be helpful for people reviewing your PR. If there are none, feel free to delete this section._

- [Jira](https://hashicorp.atlassian.net/browse/TF-8958)
- [GitHub issue](https://github.com/hashicorp/terraform-provider-tfe/issues/1030)

## Output from acceptance tests

_Please run applicable acceptance tests locally and include the output here. See [testing.md](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/testing.md) to learn how to run acceptance tests._

_If you are an external contributor, your contribution(s) will first be reviewed before running them against the project's CI pipeline._

Tested against local TFE, but these `auto_destroy_at` changes are out in production TFC.
```
➜  ~/h/terraform-provider-tfe git:(notchairmk/workspace-auto-destroy) TFE_HOSTNAME=<host> TFE_TOKEN=<token> TESTARGS="-run TestAccTFEWorkspace.*AutoDestroyAt" make testacc
TF_ACC=1 TF_LOG_SDK_PROTO=OFF go test $(go list ./... |grep -v 'vendor') -v -run TestAccTFEWorkspace.*AutoDestroyAt -timeout 15m
?   	github.com/hashicorp/terraform-provider-tfe	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-tfe/internal/client	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-tfe/internal/logging	(cached) [no tests to run]
?   	github.com/hashicorp/terraform-provider-tfe/internal/provider/validators	[no test files]
?   	github.com/hashicorp/terraform-provider-tfe/version	[no test files]
=== RUN   TestAccTFEWorkspaceDataSource_readAutoDestroyAt
--- PASS: TestAccTFEWorkspaceDataSource_readAutoDestroyAt (10.10s)
=== RUN   TestAccTFEWorkspace_createWithAutoDestroyAt
--- PASS: TestAccTFEWorkspace_createWithAutoDestroyAt (5.61s)
=== RUN   TestAccTFEWorkspace_updateWithAutoDestroyAt
--- PASS: TestAccTFEWorkspace_updateWithAutoDestroyAt (11.59s)
PASS
ok  	github.com/hashicorp/terraform-provider-tfe/internal/provider	27.816s
```
